### PR TITLE
[REVIEW] Separate column_factories into detail and cudf namespaces.

### PR DESCRIPTION
**Pull request by thomcom**
This commit puts all the templated `column_factories.hpp` functions into the `detail` namespace.

### CODE REVIEW

File: cpp/include/cudf/column/column_factories.hpp [hunk 1]
```diff
@@ -18,7 +18,6 @@
 #include <cudf/null_mask.hpp>
 #include <cudf/types.hpp>
 #include <cudf/utilities/error.hpp>
```
Comment (jrhemstad) [inline cpp/include/cudf/column/column_factories.hpp]: This should also be removed.
```suggestion
```

File: cpp/include/cudf/column/column_factories.hpp [hunk 2]
```diff
@@ -18,7 +18,6 @@
...
-#include <cudf/utilities/traits.hpp>
 #include "column.hpp"
 
 #include <rmm/thrust_rmm_allocator.h>
```
Comment (jrhemstad) [inline cpp/include/cudf/column/column_factories.hpp]: And this. 
```suggestion
```

Comment (thomcom) [reply cpp/include/cudf/column/column_factories.hpp]: 👍

**Comment by thomcom** [comment]
This PR was created because `column_factories.hpp` depends on `libcudacxx` indirectly, so third party libraries can't use it. @harrism suggested separating out the templated functions into the `detail` namespace, which is performed here. Future users of `libcudf++` might want the templated functions anyway, however.

It might be better to wait for #3507, which installs the `libcudacxx` headers during `cudf` installation.

I've made this PR available in case this functionality is desirable in addition to the installation provided in #3507.

File: cpp/include/cudf/column/column_factories.hpp (hunk 2 above)
Comment (thomcom) [reply cpp/include/cudf/column/column_factories.hpp]: Actually removing this one caused a build error from `rmm`.

Comment (harrism) [reply cpp/include/cudf/column/column_factories.hpp]: What error?

Comment (thomcom) [reply cpp/include/cudf/column/column_factories.hpp]:
```
[  8%] Building CXX object CMakeFiles/cudf.dir/src/transform/transform.cpp.o
In file included from /home/nfs/tcomer/NVIDIA/cudf/cpp/include/cudf/detail/column_factories.hpp:22:0,
                 from /home/nfs/tcomer/NVIDIA/cudf/cpp/src/transform/transform.cpp:18:
/home/nfs/tcomer/NVIDIA/cudf/cpp/include/cudf/detail/../column/column_factories.hpp:130:16: error: ‘device_vector’ in namespace ‘rmm’ does not name a template type
     const rmm::device_vector<thrust::pair<const char*, size_type>>& strings,
                ^
/home/nfs/tcomer/NVIDIA/cudf/cpp/include/cudf/detail/../column/column_factories.hpp:130:29: error: expected ‘,’ or ‘...’ before ‘<’ token
     const rmm::device_vector<thrust::pair<const char*, size_type>>& strings,
                             ^
/home/nfs/tcomer/NVIDIA/cudf/cpp/include/cudf/detail/../column/column_factories.hpp:168:16: error: ‘device_vector’ in namespace ‘rmm’ does not name a template type
     const rmm::device_vector<char>& strings,
                ^
/home/nfs/tcomer/NVIDIA/cudf/cpp/include/cudf/detail/../column/column_factories.hpp:168:29: error: expected ‘,’ or ‘...’ before ‘<’ token
     const rmm::device_vector<char>& strings,
                             ^
CMakeFiles/cudf.dir/build.make:504: recipe for target 'CMakeFiles/cudf.dir/src/transform/transform.cpp.o' failed
make[2]: *** [CMakeFiles/cudf.dir/src/transform/transform.cpp.o] Error 1
```

Comment (harrism) [reply cpp/include/cudf/column/column_factories.hpp]: Does it work if you forward declare device vector?

**Comment by harrism** [comment]
@thomcom any plans for this?

**Comment by karthikeyann** [comment]
moved to branch-0.13 
resolved conflicts, fixed compilation issues(needs second look), unit tests pass.

File: cpp/src/binaryop/compiled/binary_ops.cu [hunk 3]
```diff
@@ -14,7 +14,8 @@
  * limitations under the License.
  */
 
-#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/column_factories.hpp>
+//#include <cudf/column/column_factories.hpp>
```
Comment (karthikeyann) [inline cpp/src/binaryop/compiled/binary_ops.cu]:
```suggestion
```

File: cpp/src/groupby/sort/group_single_pass_reduction_util.cuh [hunk 4]
```diff
@@ -26,6 +27,7 @@
 #include <cudf/detail/valid_if.cuh>
 #include <cudf/types.hpp>
 
+//#include <rmm/device_buffer.hpp>
```
Comment (karthikeyann) [inline cpp/src/groupby/sort/group_single_pass_reduction_util.cuh]:
```suggestion
```

File: cpp/include/cudf/detail/column_factories.hpp [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+ * @param[in] type The desired numeric element type
+ * @param[in] size The number of elements in the column
+ * @param[in] null_mask Null mask to use for this column.
+ * @param[in] null_count Optional number of nulls in the null_mask.
+ * @param[in] stream Optional stream on which to issue all memory allocation and
+ * device kernels
+ * @param[in] mr Optional resource to use for device memory
+ * allocation of the column's `data` and `null_mask`.
+ */
+template <typename B>
```
Comment (karthikeyann) [inline cpp/include/cudf/detail/column_factories.hpp]: What is the reason to make this a templated function?

File: cpp/include/cudf/detail/column_factories.hpp [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+std::unique_ptr<column> make_timestamp_column(
```
Comment (karthikeyann) [inline cpp/include/cudf/detail/column_factories.hpp]: Does it make sense to, 
keep a declaration of template specilization `B=rmm::device_buffer` in `include/cudf/column/column_factories.hpp` 
and
keep a definition in a `.cu` file?

**Comment by thomcom** [comment]
We've worked around the problem in `cuspatial` by adding `thirdparty` submodules to match `cudf`. I don't think that this work is high priority leading up to GTC. It is also many commits behind - I'm going to close this and reopen it or recreate it if it becomes important again.

**Pull request closed without merging**